### DISCUSSION TITLE
`texlive`: `tlmgr` doesn't work

### DISCUSSION THREAD

**Discussion by ryofurue** [Homebrew/discussions#4226] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 17c872fb5275d87922a56416587cb439a5064354
Last commit: 18 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 662d1bc97c8dca79f995447cf26a0e510b5da5dc
Core tap last commit: 4 hours ago
Core tap branch: master
Core tap JSON: 17 Feb 06:51 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_blizzard_avalanche
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

As discussed here
https://tex.stackexchange.com/questions/674793/texlive-tlmgr-update-whats-the-best-configuration

the `tlmgr` command in the `texlive` formula doesn't work: `sudo tlmgr update --self` fails as described above.

If you install `texlive` using the `.pkg` package downloaded from the mactex homepage (https://www.tug.org/mactex/mactex-download.html ), the `tlmgr` command works.

By the way, I'm new here. Is this the right place to report a bug in a formula?  I've found three github projects about homebrew, but I don't know which one, if it's one of them, I should go.

**Comment by SMillerDev** [comment]
> the tlmgr command in the texlive formula doesn't work: sudo tlmgr update --self fails as described above.

No, many package managers, Homebrew included, block packages from updating themselves because they or others might stop working when you do since we can't test the compatibility.

---
> By the way, I'm new here. Is this the right place to report a bug in a formula? I've found three github projects about homebrew, but I don't know which one, if it's one of them, I should go.

For general issues this is the right place. https://docs.brew.sh/Troubleshooting has an explanation where to log issues.

**Reply by tddschn** [reply]
Could you tell me how Homebrew blocked the use of tlmgr tool?  
The source of the formula is [here](https://github.com/Homebrew/homebrew-core/blob/master/Formula/texlive.rb), I couldn't figure out how the blocking is implemented.

**Comment by ryofurue** [comment]
@SMillerDev   Thank you for the response!

> many package managers, Homebrew included, block packages from updating themselves

Is that really a conscious decision by the maintainer to "block" updates for `texlive`?

If you installed anaconda using homebrew, the `conda` command works.  It installs and updates packages .

`tlmgr` is the same as `conda`.

> because they or others might stop working when you do since we can't test the compatibility.

Sorry I don't quite understand this. You can use `tlmgr` if `texlive` is downloaded directly from mactex's homepage.  When you use `tlmgr`, your `texlive` installation may get broken.  That's the risk you agree to take.  Then, why does the maintainer of the `texlive` formula have to "block" the user when she or he agrees to take the risk?

**Reply by SMillerDev** [reply]
> Is that really a conscious decision by the maintainer to "block" updates for texlive?

Yes, https://github.com/Homebrew/homebrew-core/pull/83738 has the discussions about that.

> If you installed anaconda using homebrew, the conda command works. It installs and updates packages .

Anaconda is a cask so it only runs the upstream installer. There are also no packages that depend on it for that reason. Texlive is a formula, which is built from source and has dependents.

> You can use tlmgr if texlive is downloaded directly from mactex's homepage. 

In which case nothing depends on it.

> Then, why does the maintainer of the texlive formula have to "block" the user when she or he agrees to take the risk?

Because if another software package depends on texlive it might stop working when you do, which is then a problem for Homebrew maintainers because people will make issues for it. It's simply impossible to maintain for Homebrew with updating turned on.

**Comment by ryofurue** [comment]
> Anaconda is a cask so it only runs the upstream installer. There are also no packages that depend on it for that reason. 

I see!  That's the difference.  Thank you for the explanation. 

But that makes me wonder why texlive has to be a formula.  I guess there are distinct advantages for texlive to be a formula?  Perhaps there are already formulae that depend on texlive?

**Reply by SMillerDev** [reply]
> But that makes me wonder why texlive has to be a formula.

If something is an open source command line tool it should be a formula. Some software exists as both for various reason. For Tailscale for example the upstream packages and the command line tool differ a lot. And I believe for texlive it was made a cask for being "too difficult" and later reintroduced because people on Linux needed it (and casks are only for macOS)

**Comment by ryofurue** [comment]
Sorry for my ignorance. If I understand what you say correctly,

1. Homebrew is used on Linux (I didn't know that!!!) and the same texlive formula is used both on macOS and on Linux. Therefore, the formula version of texlive is necessary.
2. miniconda is a cask, only because it's too difficult to turn it to a formula.
3. A formula is always better than a cask because a formula can be depended on by other formulae.

> If something is an open source command line tool it should be a formula

Sorry I don't understand this "should".  Is it item 1 or 3 above?  But why "command line tool"?

> For Tailscale for example the upstream packages and the command line tool differ a lot. 

But, for texlive, there is no difference (as far as I can tell) apart from the process of installation and the availability of `tlmgr`.  Because it's much faster to install texlive with homebrew, I initially tried it, but the lack of `tlmgr` means that I have to use the upstream version.  The upstream package is updated only once a year, which is fine thanks to `tlmgr`.

**Reply by SMillerDev** [reply]
> Sorry I don't understand this "should". Is it item 1 or 3 above?

It's one of the reasons to reject a cask: https://docs.brew.sh/Acceptable-Casks#rejected-casks

> But why "command line tool"?

Firefox is open source but we don't want to build that from source.

> But, for texlive, there is no difference (as far as I can tell) apart from the process of installation and the availability of tlmgr.

So there is no reason for it to be available as a cask AND a formula

**Comment by tddschn** [comment]
Is there a work-around to make brewed texlive's tlmgr work? I accept the risk of breaking my homebrew / texlive installation.  
I really need to install some packages with it and doesn't have time to download and install texlive from mactex's website😂

**Reply by ryofurue** [reply]
I meant to reply to your message, but I wrote a separate comment.  You can use the `mactex` package.

**Comment by ryofurue** [comment]
Shortly after this discussion, I realized that there is another package of texlive called `mactex`.  (This one is a cask.). So, I recommend removing the `texlive` package and installing the `mactex` package.  Then, the `tlmgr` command will work under sudo.

**Reply by tddschn** [reply]
Thank you! I'll try mactex. I hope installing it won't take too long on my intel Mac😅

**Reply by tddschn** [reply]
@SMillerDev I wish there's a warning message about the broken tlmgr when users attempt to install texlive, and suggest to install mactex instead. What do you think?

**Reply by SMillerDev** [reply]
It's not broken, it's disabled like it would be in most package managers. Mactex isn't really the same software either so I wouldn't feel comfortable pointing users to it. But a message about `tlmgr` being disabled would be fine.

**Reply by ryofurue** [reply]
> Mactex isn't really the same software

From the user's point of view, there is no discernible difference (except for the functionality of `tlmgr`), as far as I can tell.  Because the proposed warning message is *to help the user,* it is very appropriate to point to mactex, I think.

**Reply by SMillerDev** [reply]
https://en.wikipedia.org/wiki/MacTeX

It seems they're pretty much like Google Chrome and Chromium. Just similar, not the same.

**Comment by ryofurue** [comment]
> It seems they're pretty much like Google Chrome and Chromium. Just similar, not the same.

Do *you* yourself use LaTeX as an author?  *I* use LaTeX extensively and *I* do not tell the difference.

What *practical* problems do you see if the warning mentioned mactex?  The warning would be kind to the user and there wouldn't be any real drawback.  What's the reason for the objection?

**Reply by SMillerDev** [reply]
It's redirecting users from an open source Homebrew-built package to an external package that doesn't have to follow Homebrew standards or policies, all because of a standard practice in package management.

**Reply by ryofurue** [reply]
> It's redirecting . . .

Would there be any **practical** harm in suggesting the `mactex` cask to the user?   What harm would it cause?  If it would be "redirecting users from an open source Homebrew-built package to an external package that doesn't have to follow Homebrew standards or policies", *so what?*  Who would suffer from what?

If I recommend  the`mactex` cask to my colleagues and friends, you will hate me because I'm "redirecting users from an open source Homebrew-built package to an external package that doesn't have to follow Homebrew standards or policies, all because of a standard practice in package management".  You must be hating me because I'm recommeding the "wrong" package to them.  You think they and we all **should** be using `texlive` instead of `mactex`.

**If** that's not your position, then please tell me what's wrong in suggesting the `mactex` cask to the user.

In short,  What about saying this: "tlmgr is disabled for this package. If you need it, the mactex cask may work for you."  What would go wrong?  What harm would it do? to whom? in what way?

**Reply by SMillerDev** [reply]
> Would there be any practical harm in suggesting the mactex cask to the user? 

There could be, there is much less validation for casks like mactex then there is for formulae like texlive. It could be renamed and installing malware.

> If I recommend themactex cask to my colleagues and friends, you will hate me because

No, I don't really care. But if people ask you for texlive, telling them to use mactex is not what they ask for.

**Reply by ywwry66** [reply]
> It's redirecting users from an open source Homebrew-built package to an external package that doesn't have to follow Homebrew standards or policies, all because of a standard practice in package management.

Just to add some context, Arch Linux, (whose packaging strategy is somewhat similar to Homebrew's IMO), has this [wiki page](https://wiki.archlinux.org/title/TeX_Live) "redirecting" some advanced users to install native TeX Live through CTAN other than their own packaged version, to quote
> The TeX Live snapshot packaged in the Arch Linux repositories should be fine for most users, especially if the goal is just to compile documents created by others. If, on the other hand, the user's goal is to create documents on their own; a native TeX Live installation could well turn out to be the most pain-free choice.

I think this kind of clarification is very helpful for users like @ryofurue and myself, while the risk here is really minimal.

**Reply by tddschn** [reply]
Thank you @ryofurue for the Arch wiki link! Very informative.
